cms/search/signals.py

```diff
@@ -0,0 +1,46 @@
+from django.apps import apps
+from django.db.models.signals import post_delete
+from django.dispatch import receiver
+from wagtail.models import Page
+from wagtail.signals import page_published, page_unpublished
+
+from cms.settings.base import SEARCH_INDEX_EXCLUDED_PAGE_TYPES
+
+from . import get_publisher
+
+publisher = get_publisher()
+
+
+@receiver(page_published)
+def on_page_published(sender: "Page", instance: "Page", **kwargs: dict) -> None:  # pylint: disable=unused-argument
+    """Called whenever a Wagtail Page is published (UI or code).
+    instance is the published Page object.
+    """
+    if instance.__class__.__name__ not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES:
+        publisher.publish_created_or_updated(instance)
+
+
+@receiver(page_unpublished)
+def on_page_unpublished(sender: "Page", instance: "Page", **kwargs: dict) -> None:  # pylint: disable=unused-argument
+    """Called whenever a Wagtail Page is unpublished (UI or code).
+    instance is the unpublished Page object.
+    """
+    if instance.__class__.__name__ not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES:
+        publisher.publish_deleted(instance)
+
+
+def on_page_deleted(sender: "Page", instance: "Page", **kwargs: dict) -> None:  # pylint: disable=unused-argument
```

Comment:
```suggestion
@receiver(post_delete, sender=Page)
def on_page_deleted(sender: "Page", instance: "Page", **kwargs: dict) -> None:  # pylint: disable=unused-argument
```

then you don't need `connect_page_delete_signal`

Reply:
https://docs.djangoproject.com/en/5.1/topics/signals/#connecting-to-signals-sent-by-specific-senders

same for the other two receivers

Reply:
```
@receiver(post_delete, sender=Page)
def on_page_deleted(sender: "Page", instance: "Page", **kwargs: dict) -> None:  # pylint: disable=unused-argument
    """Catches all subclass deletions of Wagtail's Page model.
    Only fires if the page is not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES.
    """
    # Only proceed if `sender` is a subclass of Wagtail Page
    if instance.specific_class.__name__ not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES:
        publisher.publish_deleted(instance)
```

does the job

Reply:
https://onswebsite.slack.com/archives/C07TZEU09FS/p1743002471280039

The core issue is that using a global `post_delete receiver` (i.e. without sender=...) has two side‐effects:

It intercepts deletions for `all models`, including `User`, which, breaks the `database‐router test` by sending the `M2M` deletes to the wrong `DB`.

When I try to narrow it with @receiver(post_delete, sender=Page), Django will only fire the signal when the exact sender is` Page`, not when `sender` is a `subclass` of `Page` like `HomePage`. Thus `page` deletion test fails.

Django will only call `on_page_deleted` if the literal sender in `post_delete.send(sender=..., ...)` is `Page`. Subclasses like` HomePage`, `ArticlePage`, etc. will not trigger it.

**Problem**: I want to fire `page deleted` logic for all `Wagtail Page subclasses` but not for every `model` in the project. Using a global `post_delete` with an issubclass(...) check changes Django’s internal DB routing for `User ` `(breaking the M2M‐delete test)`, while using `sender=Page` does not catch `HomePage` and other `subclasses`.

**Solution I found**: Dynamically connect `post_delete` to each concrete subclass of Page (and to Page itself). This way we skip subclasses nor inadvertently register for non‐Page models like User.

https://github.com/ONSdigital/dis-wagtail/pull/115/commits/d58de837170623f5774018d089580c5529a384fc

Reply:
Try the snippet I shared please

Reply:
I am so code blind.

I kept trying, only now when I copy pasted your snippet I realised I was doing `if instance.__class__.__name__ not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES:` not `if instance.specific_class.__name__ not in SEARCH_INDEX_EXCLUDED_PAGE_TYPES:`.

Mistake was myself using `instance.__class__.__name__` and not using `instance.specific_class.__name__`.

Recievers and tests updated =)

This works for `post_delete`. Adding `sender=Page` to the other published and unpublished receivers doesn't let any events fire.

Thanks @zerolab